### PR TITLE
Add 'Nullable' option to the build property page

### LINKED ISSUE #4058: Add support for Nullable build setting
This is a tracking work item related to the "nullable reference types" feature in C# 8.0 (for dev16).
We'll add design notes here as it crystallizes (is it a property or some other form of API?)
Tagging @davkean

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #4058.

This option is only available when the language version is 8.0 or greater.

---

### netcoreapp3.1 (LangVersion 8)

image

The dropdown has the following options:

image

---

### net48 (LangVersion 7.3)

The nullable control is disabled:

image

### CODE REVIEW

**Comment by davkean** [comment]
I think we should consider coming up with a way for this property to visible but disabled when changing the TFM will cause it to be enabled.

- What happens if I choose the Application tab, change from NET Core 3.0 and then choose this tab?

**Comment by davkean** [comment]
This is the help for that page: https://docs.microsoft.com/en-us/visualstudio/ide/reference/build-page-project-designer-csharp?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev16.query%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(cs.ProjectPropertiesBuild)%26rd%3Dtrue&view=vs-2019. We should document it there.

**Comment by drewnoakes** [comment]
Changing it to disable rather than hide is straightforward to do, but I felt it would make the UI more confusing for users.

Currently the state of this UI is not updated based on changes to TFM (via either UI or file on disk). The page needs to be reloaded. I looked at `PropertyControlData` for hints on how to respond to such changes but didn't see anything.

We could justify sticking help links on every UI element on the page. Does this one deserve one just because it's new?

Once the UI here is finalised I'll update the docs.

**Comment by davkean** [comment]
We don't hide elements in other cases where changing the TFM would enable it - for example, Prefer 32-bit. Can you see how it reacts when you change from 4.0 -> 4.5 in legacy?

**Comment by davidwengier** [comment]
> Does this one deserve one just because it's new?

I think this one deserves something because "Enable", "Warnings" and "Annotations" don't mean much on their own.

**Comment by drewnoakes** [comment]
Updated to respond to changes to `LangVersion`/`TargetFramework`, either through the UI or project file.

Also changed to disable rather than hide the UI.

**Comment by drewnoakes** [comment]
> I think this one deserves something because "Enable", "Warnings" and "Annotations" don't mean much on their own.

I think you could argue this about any of the options you weren't familiar with. For example, what does the `TRACE` constant mean? What's a constant? What does warning level '4' mean?

The property pages support <kbd>F1</kbd> help, though it's not very discoverable and is currently broken (I filed #5815).

**Comment by drewnoakes** [comment]
Documentation updated in https://github.com/MicrosoftDocs/visualstudio-docs-pr/pull/6024.

File: src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx [hunk 1]
```diff
@@ -1970,6 +1970,18 @@ Do you want to update the value in the .settings file?</value>
...
+  </data>
+  <data name="PPG_BuildSettings_Nullable_Enable" xml:space="preserve">
+    <value>Enable</value>
+  </data>
+  <data name="PPG_BuildSettings_Nullable_Warnings" xml:space="preserve">
+    <value>Warnings</value>
+  </data>
+  <data name="PPG_BuildSettings_Nullable_Annotations" xml:space="preserve">
+    <value>Annotations</value>
+  </data>
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx]: By placing these values in .resx file we're choosing to localize them. Is that what you want, or would it make more sense to simply use the literal values understood by the compiler?

In the case of the latter you would hard-code them in the source and use the lower-case forms.

**Comment by davkean** [comment]
The "Advanced Build Options" page is also configuration-dependent. I think you can argue this for lots of options on that page, unsafe code, Prefer 32-bit, etc.

File: src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx]: The drop down should be localized, what we persist in the project file shouldn't be.

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx]: I considered using the compiler's strings and ended up following the example set by the advanced build properties page (where I initially added this new UI before moving it to the regular build properties page).

image

image

Comment (davkean) [reply src/Microsoft.VisualStudio.Editors/Resources/Microsoft.VisualStudio.Editors.Designer.resx]: Yes, and these were localized in response to a bug.

**Comment by drewnoakes** [comment]
@tmeschter 

> If you want it to be configuration-independent, the best option is the Advanced Build Options dialog.

To be clear, in this PR the setting _behaves_ in a configuration-independent manner due to `HasConfigurationCondition="False"` on its data source.

The question is whether this behaviour is surprising to the user or not.

@davkean @davidwengier and I discussed placement on Monday and decided the setting was not really 'advanced', and that it deserved a more prominent position in the UI. So I moved it from the advanced build prop page to the regular build prop page.

But looking at it now I see that every other property on that page is actually configuration-dependent.

I agree that nullable shouldn't be configuration-dependent. It's a shame to hide the option further down, but on balance I think you're right so will move it back 🤦‍♂

**Comment by drewnoakes** [comment]
_Actually_ it turns out that every setting on the _advanced_ property page is configuration dependent too. So now I'm less sure that moving it does anything useful.

When language version was editable, it would have been the exception on all these pages.

**Comment by tmeschter** [comment]
Sorry, I missed `HasConfigurationCondition="False"` on the data source--I'm so used to thinking about these pages through the lens of the old project system. :-) I agree that with that issue out of the way, the Build page is a much better choice than the Advanced dialog. Approved.

**Pull request merged**